service/history/hsm/tree.go

```diff
@@ -185,22 +203,56 @@ func (n *Node) Path() []Key {
 
 // Outputs returns all outputs produced by transitions on this tree.
 func (n *Node) Outputs() []PathAndOutputs {
+	root := n.getRoot()
+	if root.opLog == nil {
+		return nil
+	}
+
+	currentPath := n.Path()
 	var paos []PathAndOutputs
-	if len(n.cache.outputs) > 0 {
-		paos = append(paos, PathAndOutputs{Path: n.Path(), Outputs: n.cache.outputs})
+
+	isDeleted := func(path []Key) bool {
+		for _, deletedPath := range root.opLog.DeletedPaths {
+			if isPathPrefix(deletedPath, path) {
+				return true
+			}
+		}
+		return false
 	}
-	for _, child := range n.cache.children {
-		paos = append(paos, child.Outputs()...)
+
+	if !isDeleted(currentPath) {
+		var nodeTransitions []TransitionOutputWithCount
+		for _, t := range root.opLog.Transitions {
+			if reflect.DeepEqual(t.NodePath, currentPath) {
+				nodeTransitions = append(nodeTransitions, t)
+			}
+		}
+		if len(nodeTransitions) > 0 {
+			paos = append(paos, PathAndOutputs{
+				Path:    currentPath,
+				Outputs: nodeTransitions,
+			})
+		}
+
+		for _, child := range n.cache.children {
+			paos = append(paos, child.Outputs()...)
```

Comment:
Not really seeing a great reason to call this recursively. I would do this:
1. Get the opLog from the root
2. Filter out operations if they:
    - don't start with the current node's path
    - refer to deleted paths (including transitions and deletions and if any parent up the chain was deleted)

Reply:
Agreed -- changed